movement/watch_faces/complication/activity_face.c

```diff
@@ -0,0 +1,714 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2023 Gabor L Ugray
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* ** TODO
+ * ===================
+ * -- Reset length limit to 5 min :)
+ * -- Additional power-saving optimizations
+ * OK Save logged activity; restore real length limit
+ * OK FULL behavior
+ * OK Clear blinking confirmation
+ * OK Actually clear
+ * OK Actually chirp
+ * OK Display countdown while chirping
+ * OK Quit chirp by mode press
+ * OK Update demo face with correct prefix
+ * OK Run on 1-second tick
+ * OK Fast blink for CLEAR confirm
+ * OK Low energy mode > movement_le_inactivity_deadlines
+ * -- Clear confirm: faster animation
+ * -- Write in-comment documentation
+ * OK Stop at 8 hours
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include "activity_face.h"
+#include "chirpy_tx.h"
+#include "watch.h"
+#include "watch_utility.h"
+
+// ===========================================================================
+// This part is configurable: you can edit values here to customize you activity face
+// In particular, with num_enabled_activities and enabled_activities you can choose a subset of the
+//   activities that you want to see in your watch.
+// You can also add new items to activity_names, but don't redefine or remove existing ones.
+
+// If a logged activity is shorter than this, then it won't be added to log when it ends.
+// This way scarce log slots are not taken up by aborted events that weren't real activities.
+static const uint16_t activity_min_length_sec = 10;  // 300
+
+// Supported activities. ID of activity is index in this buffer
+// W e should never change order or redefine items, only add new items when needed.
+static const char activity_names[][14] = {
```

Comment:
this should be:
`static const char activity_names[][6] = {`

Reply:
actually, make that 6 a 7 to leave room for the terminating NULL. `watch_display_string()` will stop printing after reaching the end of the display, but sprintf() will happily copy bytes into buffers until reaching a NULL.

Reply:
Silly mistake, well spotted, thanks! Fixing in next commit.